pkg/chunk/local/boltdb_index_client.go

```diff
@@ -117,41 +119,45 @@ func (b *boltIndexClient) Stop() {
 	b.wait.Wait()
 }
 
-func (b *boltIndexClient) NewWriteBatch() chunk.WriteBatch {
+func (b *BoltIndexClient) NewWriteBatch() chunk.WriteBatch {
 	return &boltWriteBatch{
 		tables: map[string]map[string][]byte{},
 	}
 }
 
-func (b *boltIndexClient) getDB(name string) (*bbolt.DB, error) {
+func (b *BoltIndexClient) GetDB(name string, operation int) (*bbolt.DB, error) {
 	b.dbsMtx.RLock()
 	db, ok := b.dbs[name]
 	b.dbsMtx.RUnlock()
 	if ok {
 		return db, nil
 	}
 
-	b.dbsMtx.Lock()
-	defer b.dbsMtx.Unlock()
-	db, ok = b.dbs[name]
-	if ok {
-		return db, nil
-	}
+	if _, err := os.Stat(path.Join(b.cfg.Directory, name)); err == nil || operation == DBOperationWrite {
```

Comment:
If `err != nil`, it's ignored. I also find this `if` statement quite unclear. I would suggest to improve it and add a comment to explain that if the DB is in read-only mode and the DB doesn't exist yet, then we should not create a new one. An example of an improved logic may be reverse the `if` logic:

```
if operation == DBOperationRead {
  if _, err := os.Stat(path.Join(b.cfg.Directory, name)); err != nil {
    return nil, <an error here>
  }
}
```